File: MyScript/edk/java/examples/inksearch/StructuredInputInkIndexing.java
package inksearch;

import java.io.File;
import java.io.IOException;

import certificates.MyCertificate;

import com.myscript.engine.Engine;
import com.myscript.engine.EngineObject;
import com.myscript.engine.FloatStructuredInput;
import com.myscript.engine.InputUnit;
import com.myscript.inksearch.Index;
import com.myscript.inksearch.StructuredInputIndexer;
import com.myscript.text.Resource;
import common.CFG;

/**
 * The <code>StructuredInputInkIndexing</code> example code shows you how to
 * perform ink indexing on a structured input.
 */
public final class StructuredInputInkIndexing
{
    /**
     * The file where the index will be stored.
     */
    private static final String INDEX_FILE = CFG.OUTPUT_DIR + "/InkIndexing-output.index";

    /**
     * Main entry point for the <code>StructuredInputInkIndexing</code> example
     * program.
     *
     * @param args command line arguments, ignored.
     */
    public static void main(String[] args) throws IOException
    {
      if( INDEX_FILE == null || INDEX_FILE.length() == 0 )
      {
        System.err.println("/!\\ please choose a valid filename for index file. You can choose index filename by editing INDEX_FILE static member variable of StructuredInputInkIndexing class");
        System.exit(-1);
      }

      System.out.println("1) creating a new Engine initialized with your certificate");
      final Engine engine = Engine.create(MyCertificate.getBytes());
      System.out.println(" . Engine created successfully");
      System.out.println();

      System.out.println("2) creating a new StructuredInputIndexer");
      final StructuredInputIndexer indexer = StructuredInputIndexer.create(engine);
      System.out.println(" . StructuredInputIndexer created successfully");
      System.out.println();

      System.out.println("3) loading and attaching resources to the indexer");
      final Resource ak = (Resource) EngineObject.load(engine, CFG.getAkRes());
      System.out.println(" . AlphabetKnowledge resource loaded successfully");
      indexer.attach(ak);
      System.out.println(" . AlphabetKnowledge resource attached successfully");
      System.out.println(" . disposing of the AlphabetKnowledge resource (not needed anymore)");
      ak.dispose();
      System.out.println(" . AlphabetKnowledge resource disposed of successfully");

      final Resource lk = (Resource)EngineObject.load(engine, CFG.getLkTextRes());
      System.out.println(" . Main LinguisticKnowledge resource loaded successfully");
      indexer.attach(lk);
      System.out.println(" . Main LinguisticKnowledge resource attached successfully");
      System.out.println(" . disposing of the main LinguisticKnowledge resource (not needed anymore)");
      lk.dispose();
      System.out.println(" . Main LinguisticKnowledge resource disposed of successfully");
      if (!CFG.getLang().contains("en"))
      {
        final String path;

        if (CFG.getLang().contains("zh") || CFG.getLang().contains("ja") || CFG.getLang().contains("ko"))
          path = CFG.getLkEnglishSecondary2Res();
        else if (CFG.getLang().contains("hi"))
          path = CFG.getLkEnglishSecondary3Res();
        else
          path = CFG.getLkEnglishSecondary1Res();

        final Resource lkEnglishSecondary = (Resource)EngineObject.load(engine, path);
        System.out.println(" . Secondary English LinguisticKnowledge resource loaded successfully");
        indexer.attach(lkEnglishSecondary);
        System.out.println(" . Secondary English LinguisticKnowledge resource attached successfully");
        System.out.println(" . disposing of the secondary LinguisticKnowledge resource (not needed anymore)");
        lkEnglishSecondary.dispose();
        System.out.println(" . Secondary LinguisticKnowledge resource disposed of successfully");
      }
      System.out.println();
      
      System.out.println("4) creating a new FloatStructuredInput");
      final FloatStructuredInput input = FloatStructuredInput.create(engine);
      System.out.println(" . FloatStructuredInput created successfully");
      System.out.println();

      System.out.println("5) fueling the input with strokes");
      input.startInputUnit(InputUnit.SINGLE_LINE_TEXT);
      input.addStroke(x0, y0);
      input.addStroke(x1, y1);
      input.addStroke(x2, y2);
      input.addStroke(x3, y3);
      input.addStroke(x4, y4);
      input.addStroke(x5, y5);
      input.addStroke(x6, y6);
      input.addString(" Objects");
      input.endInputUnit(InputUnit.SINGLE_LINE_TEXT);
      System.out.println(" . input fueled successfully");
      System.out.println();

      System.out.println("6) setting the input as the source of the indexer");
      indexer.setSource(input);
      System.out.println(" . input successfully set as the indexer's source");
      System.out.println();

      System.out.println("7) running the indexing process");
      indexer.run();
      System.out.println(" . indexing performed successfully");
      System.out.println();

      System.out.println("8) retrieving the Index");
      final Index index = indexer.getResult();
      System.out.println(" . Index retrieved successfully");
      System.out.println();

      System.out.println("9) storing the Index in a file for later search");
      index.store(INDEX_FILE);
      System.out.println(" . " + new File(INDEX_FILE).getCanonicalPath() + " stored successfully");
      System.out.println();

      // do not forget the 1st rule: you create it, you dispose of it!
      System.out.println("10) cleaning-up: disposing of objects and disposing of the Engine");
      index.dispose();
      indexer.dispose();
      input.dispose();
      engine.dispose();
      System.out.println(" . clean-up done successfully");
      System.out.println();

      System.out.println("end of example:");
      System.out.println(" you just learnt how to perform ink indexing on a structured input");
    }

    /**
     * Strokes that will be indexed.
     */
    private static float[] x0 = new float[]{374, 374, 374, 374, 376, 378, 379, 381, 384, 387, 389, 390};
    private static float[] y0 = new float[]{493, 491, 493, 499, 511, 520, 522, 518, 509, 499, 493, 494};
    private static float[] x1 = new float[]{386, 388, 389, 391, 391, 391, 391, 391, 391, 394};
    private static float[] y1 = new float[]{518, 517, 516, 514, 511, 511, 513, 516, 517, 517};
    private static float[] x2 = new float[]{397, 396};
    private static float[] y2 = new float[]{502, 502};
    private static float[] x3 = new float[]{397, 398, 400, 400, 403, 404, 405, 404, 402, 400, 400, 402, 404, 406, 406, 407, 407};
    private static float[] y3 = new float[]{519, 518, 516, 514, 514, 516, 518, 520, 522, 520, 518, 516, 514, 514, 516, 518, 520};
    private static float[] x4 = new float[]{409, 409, 408};
    private static float[] y4 = new float[]{505, 503, 503};
    private static float[] x5 = new float[]{413, 414, 412, 412, 412, 414, 416, 417, 417, 415, 413};
    private static float[] y5 = new float[]{517, 516, 516, 518, 521, 522, 520, 517, 514, 513, 515};
    private static float[] x6 = new float[]{420, 420, 420, 420, 421, 422, 423, 424, 424, 425, 426, 428, 429, 430, 432, 433, 435, 435};
    private static float[] y6 = new float[]{518, 517, 515, 513, 511, 511, 511, 514, 517, 518, 518, 516, 514, 514, 515, 518, 519, 519};

} // StructuredInputInkIndexing

File: MyScript/edk/c/examples/inksearch/StructuredInputInkIndexing.c
/**
 * The `StructuredInkIndexing` example code shows you how to perform ink
 * indexing on a structured input.
 */


// -- configuration ------------------------------------------------------------

// common configuration parameters
#include "../common/config.h"

// include the MyScript Engine API header(s)
#include <c/MyScriptEngine.h>
#include <c/MyScriptInkSearch.h>

// include the certificate
#include "../certificates/MyCertificate.h"


// -- example body -------------------------------------------------------------

#include <stdio.h>

// global engine (for the sake of readability)
voEngine engine = NULL;

// common utility functions
#include "../common/common.h"

// digital ink
static float x0[] = {374, 374, 374, 374, 376, 378, 379, 381, 384, 387, 389, 390};
static float y0[] = {493, 491, 493, 499, 511, 520, 522, 518, 509, 499, 493, 494};
static float x1[] = {386, 388, 389, 391, 391, 391, 391, 391, 391, 394};
static float y1[] = {518, 517, 516, 514, 511, 511, 513, 516, 517, 517};
static float x2[] = {397, 396};
static float y2[] = {502, 502};
static float x3[] = {397, 398, 400, 400, 403, 404, 405, 404, 402, 400, 400, 402, 404, 406, 406, 407, 407};
static float y3[] = {519, 518, 516, 514, 514, 516, 518, 520, 522, 520, 518, 516, 514, 514, 516, 518, 520};
static float x4[] = {409, 409, 408};
static float y4[] = {505, 503, 503};
static float x5[] = {413, 414, 412, 412, 412, 414, 416, 417, 417, 415, 413};
static float y5[] = {517, 516, 516, 518, 521, 522, 520, 517, 514, 513, 515};
static float x6[] = {420, 420, 420, 420, 421, 422, 423, 424, 424, 425, 426, 428, 429, 430, 432, 433, 435, 435};
static float y6[] = {518, 517, 515, 513, 511, 511, 511, 514, 517, 518, 518, 516, 514, 514, 515, 518, 519, 519};

// path of the file where the index will be stored.
#define INDEX_FILE OUTPUT_DIR "/InkIndexing-output.index"

// main function
int main()
{
  voStructuredInput input = NULL;
  voString string = {" Objects", -1};
  voIndex index = NULL;
  voStructuredInputIndexer indexer = NULL;
  voResource ak = NULL;
  voResource lk = NULL;
  voResource lkEnglishSecondary = NULL;

  if (strlen(LANG) == 0 || strlen(HWSTYLE) == 0)
  {
    fprintf(stderr, "/!\\ please edit the common/config.h file and set:\n");
    fprintf(stderr, " . the LANG variable\n");
    fprintf(stderr, " . the HWSTYLE variable\n");
    fflush(stderr);
    exit(EXIT_FAILURE);
  }

  if (strlen(INDEX_FILE) == 0)
  {
    printf("/!\\ please edit this example program setting the INDEX_FILE variable to a valid file name\n");
    exit(EXIT_FAILURE);
  }
  
  // -- library initialization -------------------------------------------------
  if (!loadMyScriptEngine(MYSCRIPT_ENGINE_LIBRARY_DIR, NULL))
    return (fprintf(stderr, "failed to load the MyScriptEngine library\n"), EXIT_FAILURE);

  // -- example code -----------------------------------------------------------
  printf("1) creating a new Engine initialized with your certificate\n");
  engine = voCreateEngine(VO_MSE_VER, &myCertificate, NULL);
  verify(engine != NULL);
  printf(" . Engine created successfully\n");
  printf("\n");

  printf("2) creating a new StructuredInputIndexer\n");
  indexer = voCreateStructuredInputIndexer(engine);
  verify(indexer != NULL);
  printf(" . StructuredInputIndexer created successfully\n");
  printf("\n");

  printf("3) loading and attaching resources to the indexer\n");
  ak = loadEngineObject(AK_RES);
  printf(" . AlphabetKnowledge resource loaded successfully\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, indexer, ak));
  printf(" . AlphabetKnowledge resource attached successfully\n");
  lk = loadEngineObject(LK_TEXT_RES);
  printf(" . Main LinguisticKnowledge resource loaded successfully\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, indexer, lk));
  printf(" . Main LinguisticKnowledge resource attached successfully\n");
  if (strstr(LANG, "en") == NULL)
  {
    const char* path;
    if (strstr(LANG, "zh") != NULL || strstr(LANG, "ja") != NULL || strstr(LANG, "ko") != NULL)
      path = LK_ENGLISH_SECONDARY_2_RES;
    else if (strstr(LANG, "hi") != NULL)
      path = LK_ENGLISH_SECONDARY_3_RES;
    else
      path = LK_ENGLISH_SECONDARY_1_RES;

    lkEnglishSecondary = loadEngineObject(path);
    printf(" . Secondary English LinguisticKnowledge resource loaded successfully\n");
    verify(VO_IFACE(engine, IAttachTarget)->attach(engine, indexer, lkEnglishSecondary));
    printf(" . Secondary English LinguisticKnowledge resource attached successfully\n");
  }
  printf("\n");

  printf("4) creating an FloatStructuredInput\n");
  input = voCreateFloatStructuredInput(engine);
  verify(input != NULL);
  printf(" . FloatStructuredInput created successfully\n");
  printf("\n");

  printf("5) fueling the input with strokes\n");
  verify(VO_IFACE(engine, IFloatStructuredInput)->startInputUnit(engine, input, VO_SINGLE_LINE_TEXT));
  verify(VO_IFACE(engine, IFloatStructuredInput)->addStroke(engine, input, x0, sizeof(*x0), y0, sizeof(*y0), (int)(sizeof(x0)/sizeof(*x0))));
  verify(VO_IFACE(engine, IFloatStructuredInput)->addStroke(engine, input, x1, sizeof(*x1), y1, sizeof(*y1), (int)(sizeof(x1)/sizeof(*x1))));
  verify(VO_IFACE(engine, IFloatStructuredInput)->addStroke(engine, input, x2, sizeof(*x2), y2, sizeof(*y2), (int)(sizeof(x2)/sizeof(*x2))));
  verify(VO_IFACE(engine, IFloatStructuredInput)->addStroke(engine, input, x3, sizeof(*x3), y3, sizeof(*y3), (int)(sizeof(x3)/sizeof(*x3))));
  verify(VO_IFACE(engine, IFloatStructuredInput)->addStroke(engine, input, x4, sizeof(*x4), y4, sizeof(*y4), (int)(sizeof(x4)/sizeof(*x4))));
  verify(VO_IFACE(engine, IFloatStructuredInput)->addStroke(engine, input, x5, sizeof(*x5), y5, sizeof(*y5), (int)(sizeof(x5)/sizeof(*x5))));
  verify(VO_IFACE(engine, IFloatStructuredInput)->addStroke(engine, input, x6, sizeof(*x6), y6, sizeof(*y6), (int)(sizeof(x6)/sizeof(*x6))));
  verify(VO_IFACE(engine, IFloatStructuredInput)->addString(engine, input, NULL, &string));
  verify(VO_IFACE(engine, IFloatStructuredInput)->endInputUnit(engine, input, VO_SINGLE_LINE_TEXT));
  printf(" . input fueled successfully\n");
  printf("\n");

  printf("6) setting the input as the source of the indexer\n");
  verify(VO_IFACE(engine, IInputConsumer)->setSource(engine, indexer, input));
  printf(" . input successfully set as the indexer's source\n");
  printf("\n");

  printf("7) running the indexing process\n");
  verify(VO_IFACE(engine, IRunnable)->run(engine, indexer, NULL, NULL));
  printf(" . indexing performed successfully\n");
  printf("\n");

  printf("8) retrieving the Index\n");
  index = VO_IFACE(engine, IIndexingResultProvider)->getResult(engine, indexer);
  verify(index != NULL);
  printf(" . Index retrieved successfully\n");
  printf("\n");

  printf("9) storing the Index in a file for later search\n");
  storeEngineObject(index, INDEX_FILE);
  printf(" . Index stored successfully to: ");
  printPath(stdout, INDEX_FILE);
  printf("\n");
  printf("\n");

  printf("10) cleaning-up: releasing all created objects and destroying the Engine\n");
  verify(voReleaseObject(engine, index));
  verify(voReleaseObject(engine, input));
  verify(voReleaseObject(engine, ak));
  if (lkEnglishSecondary != NULL)
    verify(voReleaseObject(engine, lkEnglishSecondary));
  verify(voReleaseObject(engine, lk));
  verify(voReleaseObject(engine, indexer));
  verify(voDestroyEngine(engine));
  printf(" . clean-up done successfully\n");
  printf("\n");

  printf("end of example:\n");
  printf(" you just learnt how to perform ink indexing on a structured input\n");

  // -- finalization -----------------------------------------------------------
  unloadMyScriptEngine();

  return EXIT_SUCCESS;
}
